Clarify resolving implicit connections (3.1.1) (#3823)
OAI/OpenAPI-Specification: versions/3.1.1.md

When parsing an OAD, JSON or YAML objects are parsed into specific Objects (such as [Operation Objects](#operationObject), [Response Objects](#responseObject), [Reference Objects](#referenceObject), etc.) based on the parsing context.  Depending on how references are arranged, a given JSON or YAML object can be parsed in multiple different contexts:

* As a complete OpenAPI Description document
* As the Object type implied by its parent Object within the document
* As a reference target, with the Object type matching the reference source's context

If the same JSON/YAML object is parsed multiple times and the respective contexts require it to be parsed as _different_ Object types, the resulting behavior is _implementation defined_, and MAY be treated as an error if detected.  An example would be referencing an empty Schema Object under `#/components/schemas` where a Path Item Object is expected, as an empty object is valid for both types.  For maximum interoperability, it is RECOMMENDED that OpenAPI Description authors avoid such scenarios.

> #### <a name="resolvingIndirectLinkages"></a>Resolving Indirect Connections

Reviewer: Why is the anchor not the camelCasedVersion of the section title?

Author: Because I changed the title a bunch.  Good catch!

[Link Object](#linkObject) `operationId` | [Path Item Object](#pathItemObject) `operationId` | `operationRef`

A fourth implicit connection, which involves appending the templated URL paths of the [Paths Object](#pathsObject) to the appropriate [Server Object](#serverObject)'s `url` field, is unambiguous because only the entry document's Paths Object contributes URLs to the described API.

For resolving indirect connections from a referenced (non-entry) document, it is RECOMMENDED that tools support one or both of the following approaches, and it is likewise RECOMMENDED that OAD authors organize their documents to work well with them.

##### OADs Using Only Complete Documents

The approach in this section is RECOMMENDED for OADs that are composed entirely of complete OpenAPI documents, optionally including standalone JSON Schema documents as long as any Discriminator Objects within them only use explicit `mapping` URIs for all possible values.
This OAD design is RECOMMENDED when sharing components among multiple OADs.

> In this approach, each document is self-contained and resolves components to its own Components Object.

Reviewer: So I can't reference component schemas from another OAD object, I have to extract them into separate JSON Schema documents?

For whom would that be a simplification?

Author: No that's not what this says at all.  There is nothing here about forbidding Schema Objects in OpenAPI documents.  What is causing you to arrive at this interpretation?

When resolving an `operationId`, all Path Item Objects from all parsed documents are considered.
This requires ensuring that all referenced documents have been parsed prior to determining an `operationId` to be unresolvable.

**Note:** Currently, Security Requirement Objects do not support a URI-based mechanism for connecting to Security Scheme Objects, which means that in this approach, they can only connect to Security Schemes within the same document.

> This limitation is expected to be addressed in the next minor version of this specification.

Reviewer: ```suggestion
This limitation is expected to be addressed in a future version of this specification.
```
Don't over-promise.

Author: The PR for this is up for 3.2, so if we can agree on it, then "next minor" is correct and sets a better expectation.

Author: On second thought, it's better to be more vague as we can always add specifics later if we agree to the 3.2 change.  But I don't want us to have to think about 3.2 before doing these patch releases, so I'll commit this suggestion.

##### OADs With Only One OpenAPI Object

The approach in this section is RECOMMENDED for OADs where the entry OpenAPI document is the only document containing an OpenAPI Object.
This approach is most suitable for splitting an otherwise self-contained single-document OAD into multiple documents for author convenience.

> In this approach, all component names are resolved through the entry document's Components Object, and all `operationId`s are resolved through Path Item Objects reachable through the `paths` or `webhooks` fields.

Author: There is nothing here about not being able to directly reference schemas.  This is about how component names (not URIs) are resolved.  Do you understand the distinction?  If not, let's figure out better wording.
